### ISSUE TITLE
File names are repeated in Code analysis file combobox

### ISSUE THREAD

**Issue by ccordoba12** [spyder-ide/spyder#12883]
Instead of registering each file name once, we're doing it several times. This is how the problem looks graphically:

![imagen](https://user-images.githubusercontent.com/365293/83071209-5f8b6000-a032-11ea-8744-c9ebc0dc8f1e.png)

@juanis2112, please wait until @steff456 fix this one before filming the pending gif for our docs.

**Comment by goanpeca** [comment]
The original issue:
https://github.com/spyder-ide/spyder/issues/12740
and the PR https://github.com/spyder-ide/spyder/pull/1287 resulted in a mix of concepts.

The `History` option inside the options never referred to the items in the ComboBox nor to the results in the tree. First the combobox is tied to the editor, so every-time a new file it is open, it will add it to the combobox. Second, the default value was 50 (a value way too high to refer to the Combobox items), or too small to refer to the history of results.

The original history refer to the lines of output saved in the `rdata` attribute, see: https://github.com/spyder-ide/spyder/blob/master/spyder/plugins/pylint/widgets/pylintgui.py#L323

So the original issue talked about the history of the combobox, the PR now refers to the amount of results in the tree item, but the code is still using a hardcoded value of 100 to trim the results that are saved to disk. 

What do we want to change?

The maximum history of the combobox?
The maximum amount of results in the tree?
The code that saved the results to disk is still using the initial history value.

**Comment by steff456** [comment]
You are right Gonzalo! I understood that the `max_entries` was for the results of the code analysis instead of the number of entries of the combo box. I'm not sure if there's a need for trimming the results of the code analysis so we can have the changes of PR #12874 or revert them.

I think that the `max_entries` configurable option is for the history of the combobox. Even though, PR #12884 is just avoiding duplicate filenames in the combobox because the option is connected each time that the editor focus is changed, that's why we had multiple times the same file name.
So, I just need to know if I the changes of PR #12874 need a new option (or revert) and in the new one I will connect `max_entries` with the combobox

**Comment by goanpeca** [comment]
The repeated items in the combobox is actually a problem that does need fixing 👍 

Now to be honest, I am frankly not sure if having that `history` option is that useful at all for any of the current options, so I would actually vote to remove it. 

The way this plugin works is that the results only arrive after everything is finished, which for a long file might be a bit annoying. I would say that the option that made sense would be like the `Find in Files` plugin work, that is to limit the results (like what @steff456 did) but we would need to do w things:
1. Use a clear name like find in files
<img width="261" alt="Screen Shot 2020-05-27 at 18 33 58" src="https://user-images.githubusercontent.com/3627835/83082290-a3895f80-a048-11ea-95f4-5ab8112483fa.png">
2. Make this plugin be able to process results as they arrive (like find in files does it), instead of all once it is finished (which currently provides a weird UX since we have to wait for.... some time on big files with no output on what is happennng)

**Comment by ccordoba12** [comment]
> The way this plugin works is that the results only arrive after everything is finished, which for a long file might be a bit annoying

I disagree because that could hide potentially important information. The best option is for users to hide results by using the Pylint messages (that was the purpose of adding them in PR #12803).

Instead, I think what @juanis2112 and I expected (that is, that History controls the number of entries in the combobox), is what we need to implement here.

**Comment by goanpeca** [comment]
That is a valid point, but then the default value of 50 (and 100 in the init) should be lowered.

I still suggest we improve the UX with point 2 (in a Separate PR)

**Comment by ccordoba12** [comment]
> That is a valid point, but then the default value of 50 (and 100 in the init) should be lowered.

Agreed. I think a minimum of 5 and max of 20 would be fine. Do you agree?

> I still suggest we improve the UX with point 2 (in a Separate PR)

Sure, but I don't think that's possible. @CAM-Gerlach, what do you say?

**Comment by goanpeca** [comment]
20 sounds good!

**Comment by goanpeca** [comment]
It should be possible to parse the report  as output arrives from stdout

![maximize](https://user-images.githubusercontent.com/3627835/83090698-fb32c580-a05e-11ea-8bbb-52f30b358e99.gif)

Also, since pylint allows checking folders, it would probably be usefull if we allowed for files but also folders?

**Comment by ccordoba12** [comment]
> It should be possible to parse the report as output arrives from stdout

Ok, then please open a new issue about it.

> Also, since pylint allows checking folders, it would probably be usefull if we allowed for files but also folders?

That's already possible, as long as the folder is a Python module.

**Comment by goanpeca** [comment]
Ok, then please open a new issue about it.>

See https://github.com/spyder-ide/spyder/issues/12887

>That's already possible, as long as the folder is a Python module.

Indeed, but it is not very discoverable, see: https://github.com/spyder-ide/spyder/issues/12888

**Comment by CAM-Gerlach** [comment]
> Agreed. I think a minimum of 5 and max of 20 would be fine. Do you agree?

Sorry, I'm not sure what you mean by "minimum of 5 and max of 20"? Do you mean the minimum and maximum number of files shown in the combobox? Artificially limiting users to selecting between 5 and 20 for their own value? Or picking a default somewhere between those values? I think 20's okay, though I don't see a huge downside of picking a higher value...

> That's already possible, as long as the folder is a Python module.

(* package, btw, a module is an individual file while a (sub)package is a dir with an `__init__.py`)

> Sure, but I don't think that's possible. @CAM-Gerlach, what do you say?

Seems like it should theoretically be possible, assuming `QProcess.readyReadStandardOutput` calls its callback with each line or at some reasonable frequency; we should be able to run most of the current `finished` callback that parses the output and sends it to the GUI widget with each of the aforementioned calls (or at least every n of them, etc), and then just do one final refresh and any remaining cleanup on `QProcess.finished`; we'd have to ensure a reasonable update rate to balance usefullness and overhead. Seems like it should be relatively simple given the existing structure but the devil is, as always, in the details. @goanpeca you're surely much more familiar with `QProcess` objects and how that all works, and I'm not expert on the pylint plugin...thoughts?

A stopgap to improve the UX while the user waits is to render a progress indicator like was added to the Help, but if its really as relatively straightforward as it would seem, might be easier to just implement the above. What would be the target version for this?

**Comment by ccordoba12** [comment]
> Do you mean the minimum and maximum number of files shown in the combobox? Artificially limiting users to selecting between 5 and 20 for their own value?

Yes and yes.

> Or picking a default somewhere between those values

The default could be 20, but we were not talking about that.

> I think 20's okay, though I don't see a huge downside of picking a higher value...

Perhaps 30 or 50 is better? What do you think?

**Comment by CAM-Gerlach** [comment]
> > Do you mean the minimum and maximum number of files shown in the combobox? Artificially limiting users to selecting between 5 and 20 for their own value?

> Yes and yes.

I see. Is there a technical rationale for artificially locking users to these values? If not, and we're going to offer the config option anyway, what's the reasoning behind imposing these limits (at least between some set of sane values, like 1-999)? At least in my experience, they just tend to frustrate users, and in fact were a significant problem for my as I was trying to reproduce this issue ( #12740 ) to begin with.

> Perhaps 30 or 50 is better? What do you think?

30 is probably enough for a default, I'd think...better to err on the side of preserving rather than discarding past state.
